### PR TITLE
Replace use of TargetFramework in native build

### LINKED ISSUE #1348: Fix use of TargetFramework property in dnx.vcxproj
We shouldn't use the TargetFramework property any more. The split is actually by runtime, not framework. Instead of building the dnx.exe bootstrapper in "dnx451" and "dnxcore50" flavors, we should use "clr" and "coreclr_[platform]" as the names.

### PR DESCRIPTION
**Pull request by anurse**
Also updated all the project.jsons and MSBuild projects to copy binaries in to your existing DNXes during development time for ALL combinations of CoreCLR/CLR/Mono and x86/x64

/cc @davidfowl @ChengTian @GrabYourPitchforks 

Fixes #1348

### CODE REVIEW

File: src/Microsoft.Framework.CommandLineUtils/project.json [hunk 1]
```diff
@@ -1,14 +1,24 @@
...
+        "dnx451" : { },
+        "dnxcore50" : { 
+            "dependencies": {
+                "System.Collections": "4.0.10-beta-*",
+                "System.Runtime": "4.0.20-beta-*"
+            }
+        }
+   },
+
+    "scripts": {
```
Comment (anurse) [inline src/Microsoft.Framework.CommandLineUtils/project.json]: Whoops, this doesn't need to be here, I'll clean it up.

File: src/dnx.coreclr/dnx.coreclr.vcxproj [hunk 2]
```diff
@@ -1,224 +1,226 @@
...
+  <ImportGroup Label="ExtensionTargets">
+  </ImportGroup>
+  <Target Name="CopyManagedForDebuggingLocally" AfterTargets="Build" Condition="'$(Configuration)'=='Debug'">
+    <PropertyGroup>
+      <_CopyForDebuggingDest Condition="'$(Platform)'=='Win32'">$(SolutionDir)\artifacts\build\dnx-coreclr-win-x86\bin</_CopyForDebuggingDest>
+      <_CopyForDebuggingDest Condition="'$(Platform)'=='x64'">$(SolutionDir)\artifacts\build\dnx-coreclr-win-x64\bin</_CopyForDebuggingDest>
+    </PropertyGroup>
+    <Copy SourceFiles="$(SolutionDir)\bin\$(Platform)\$(Configuration)\dnx.coreclr.dll" DestinationFolder="$(_CopyForDebuggingDest)" />
+    <Copy SourceFiles="$(SolutionDir)\bin\$(Platform)\$(Configuration)\dnx.coreclr.pdb" DestinationFolder="$(_CopyForDebuggingDest)" />
+  </Target>
```
Comment (anurse) [inline src/dnx.coreclr/dnx.coreclr.vcxproj]: This target is the main new component of this change

File: src/dnx/dnx.cpp [hunk 3]
```diff
@@ -31,9 +31,9 @@ bool StringsEqual(LPCWSTR const pszStrA, LPCWSTR const pszStrB)
 
 bool StringEndsWith(LPCWSTR const pszStr, LPCWSTR const pszSuffix)
 {
-    int nStrLen = wcsnlen_s(pszStr, MAX_PATH);
-    int nSuffixLen = wcsnlen_s(pszSuffix, MAX_PATH);
-    int nOffset = nStrLen - nSuffixLen;
+    size_t nStrLen = wcsnlen_s(pszStr, MAX_PATH);
```
Comment (anurse) [inline src/dnx/dnx.cpp]: Fixes C++ compiler warnings about casting size_t to int

File: src/dnx/dnx.vcxproj [hunk 4]
```diff
@@ -1,199 +1,210 @@
...
+  <PropertyGroup Label="Globals">
+    <SolutionDir Condition="'$(SolutionDir)'==''">$([System.IO.Path]::GetFullPath("$(MSBuildProjectDirectory)\..\..\"))</SolutionDir>
+    <ProjectGuid>{D0E2FB09-0FEA-478A-9068-D6AA420C6DED}</ProjectGuid>
+    <Keyword>Win32Proj</Keyword>
+    <RootNamespace>dnx</RootNamespace>
+    <ProjectName>dnx</ProjectName>
+    <RuntimeType Condition="$(RuntimeType) == ''">clr</RuntimeType>
+    <RuntimeDefine Condition="$(RuntimeType) == 'coreclr-win'">CORECLR_WIN</RuntimeDefine>
+    <RuntimeDefine Condition="$(RuntimeType) == 'coreclr-unix'">CORECLR_UNIX</RuntimeDefine>
+    <RuntimeDefine Condition="$(RuntimeType) == 'coreclr-darwin'">CORECLR_DARWIN</RuntimeDefine>
```
Comment (anurse) [inline src/dnx/dnx.vcxproj]: Name for this? Darwin is commonly used, but I could do Mac instead.

File: src/dnx/dnx.vcxproj [hunk 5]
```diff
@@ -1,199 +1,210 @@
...
+  <ImportGroup Label="PropertySheets" Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+  </ImportGroup>
+  <ImportGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'" Label="PropertySheets">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+  </ImportGroup>
+  <PropertyGroup Label="UserMacros" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
+    <LinkIncremental>true</LinkIncremental>
+    <OutDir>$(SolutionDir)bin\$(Platform)\$(Configuration)\$(RuntimeType)\</OutDir>
```
Comment (anurse) [inline src/dnx/dnx.vcxproj]: OutDirs here (and below) now all use RuntimeType instead of TargetFramework

File: src/dnx/dnx.vcxproj [hunk 6]
```diff
@@ -1,199 +1,210 @@
...
+  <Target Name="CopyManagedForDebuggingLocally" AfterTargets="Build" Condition="'$(Configuration)'=='Debug'">
+    <PropertyGroup>
+      <_CopyForDebuggingDest Condition="'$(Platform)|$(RuntimeType)'=='Win32|clr'">$(SolutionDir)\artifacts\build\dnx-clr-win-x86\bin</_CopyForDebuggingDest>
+      <_CopyForDebuggingDest Condition="'$(Platform)|$(RuntimeType)'=='x64|clr'">$(SolutionDir)\artifacts\build\dnx-clr-win-x64\bin</_CopyForDebuggingDest>
+      <_CopyForDebuggingDest Condition="'$(Platform)|$(RuntimeType)'=='Win32|coreclr-win'">$(SolutionDir)\artifacts\build\dnx-coreclr-win-x86\bin</_CopyForDebuggingDest>
+      <_CopyForDebuggingDest Condition="'$(Platform)|$(RuntimeType)'=='x64|coreclr-win'">$(SolutionDir)\artifacts\build\dnx-coreclr-win-x64\bin</_CopyForDebuggingDest>
+    </PropertyGroup>
+    <Copy SourceFiles="$(SolutionDir)\bin\$(Platform)\$(Configuration)\$(RuntimeType)\dnx.exe" DestinationFolder="$(_CopyForDebuggingDest)" />
+    <Copy SourceFiles="$(SolutionDir)\bin\$(Platform)\$(Configuration)\$(RuntimeType)\dnx.pdb" DestinationFolder="$(_CopyForDebuggingDest)" />
+  </Target>
```
Comment (anurse) [inline src/dnx/dnx.vcxproj]: Another CopyForDebugging helper target.

**Comment by anurse** [comment]
Git didn't like the vcxproj changes so I added some comments calling out the changes.

File: src/dnx/dnx.vcxproj (hunk 4 above)
Comment (muratg) [inline src/dnx/dnx.vcxproj]: Darwin looks good.

**Comment by muratg** [comment]
Changes look good to me.  I'll get someone else with more experience on the project files to put the ship it though.

File: src/dnx/dnx.cpp [hunk 7]
```diff
@@ -31,9 +31,9 @@ bool StringsEqual(LPCWSTR const pszStrA, LPCWSTR const pszStrB)
...
+    size_t nSuffixLen = wcsnlen_s(pszSuffix, MAX_PATH);
+    size_t nOffset = nStrLen - nSuffixLen;
```
Comment (BrennanConroy) [inline src/dnx/dnx.cpp]: Making this size_t will cause the below if statement to always be false, and might cause other issues too.

**Comment by davidfowl** [comment]
Since you changed all of the project.json files could you group the scripts? dnx451 then dnxcore50

File: src/dnx/dnx.vcxproj (hunk 4 above)
Comment (davidfowl) [inline src/dnx/dnx.vcxproj]: Take out everything but windows. We won't be using vcxproj to build anything but windows

File: src/dnx/dnx.cpp (hunk 7 above)
Comment (anurse) [inline src/dnx/dnx.cpp]: :+1: I'll review the usage of these.

File: src/dnx/dnx.cpp [hunk 8]
```diff
@@ -5,395 +5,401 @@
 
 void GetModuleDirectory(HMODULE module, LPWSTR szPath)
 {
-    DWORD dirLength = GetModuleFileName(module, szPath, MAX_PATH);
-    for (dirLength--; dirLength >= 0 && szPath[dirLength] != '\\'; dirLength--);
-    szPath[dirLength + 1] = L'\0';
+	DWORD dirLength = GetModuleFileName(module, szPath, MAX_PATH);
+	for (dirLength--; dirLength >= 0 && szPath[dirLength] != '\\'; dirLength--);
+	szPath[dirLength + 1] = L'\0';
```
Comment (anurse) [inline src/dnx/dnx.cpp]: The default VS C/C++ Tabs setting is to use hard-tabs. This all appeared when I Ctrl-K Ctrl-D'd on the cpp file. If we want to change from the default, I'm fine with that, but it seems better to use the default.

File: src/dnx/dnx.cpp [hunk 9]
```diff
@@ -5,395 +5,401 @@
...
 
 bool StringEndsWith(LPCWSTR const pszStr, LPCWSTR const pszSuffix)
 {
-    int nStrLen = wcsnlen_s(pszStr, MAX_PATH);
-    int nSuffixLen = wcsnlen_s(pszSuffix, MAX_PATH);
-    int nOffset = nStrLen - nSuffixLen;
+	// Casting to int is safe here because MAX_PATH (260) always fits in an int
+	int nStrLen = static_cast<int>(wcsnlen_s(pszStr, MAX_PATH));
+	int nSuffixLen = static_cast<int>(wcsnlen_s(pszSuffix, MAX_PATH));
+	int nOffset = nStrLen - nSuffixLen;
```
Comment (anurse) [inline src/dnx/dnx.cpp]: @BrennanConroy fix for your size_t note.

File: src/dnx/dnx.vcxproj (hunk 4 above)
Comment (anurse) [inline src/dnx/dnx.vcxproj]: Good point :)

File: src/dnx/dnx.cpp (hunk 8 above)
Comment (davidfowl) [inline src/dnx/dnx.cpp]: No tabs :smile:

File: src/dnx/dnx.cpp (hunk 9 above)
Comment (BrennanConroy) [inline src/dnx/dnx.cpp]: :+1:

**Comment by anurse** [comment]
Shelving this. There are some issues with the 'def' file and this is mostly cosmetic right now. I'm going to return to it when I do some native bootstrapper changes to support better servicing of DNX runtimes.

**Pull request closed without merging**